File: test/napi/lib/boxed.js
const addon = require("..");
const { expect } = require("chai");
const assert = require("chai").assert;

class Person {
  constructor(name) {
    this._person = addon.person_new(name);
  }

  greet() {
    return addon.person_greet(this._person);
  }
}

class RefPerson {
  constructor(name) {
    this._person = addon.ref_person_new(name);
  }

  greet() {
    return addon.ref_person_greet(this._person);
  }

  setName(name) {
    addon.ref_person_set_name(this._person, name);

    return this;
  }

  fail() {
    addon.ref_person_fail(this._person);
  }
}

describe("boxed", function () {
  it("can call methods", function () {
    const person = new Person("World");
    const greeting = person.greet();

    assert.strictEqual(greeting, "Hello, World!");
  });

  it("can call methods wrapped in a RefCell", function () {
    const person = new RefPerson("World");
    const greeting = person.greet();

    assert.strictEqual(greeting, "Hello, World!");
  });

  it("can mutate with ref cell", function () {
    const person = new RefPerson("World").setName("Universe");
    const greeting = person.greet();

    assert.strictEqual(greeting, "Hello, Universe!");
  });

  it("should dynamically check borrowing rules", function () {
    assert.throws(() => new RefPerson("World").fail(), /already borrowed/);
  });

  it("should type check externals", function () {
    // `any::type_name` does not guarantee exact format
    // failed downcast to neon::types::boxed::JsBox<napi::js::boxed::Person>
    assert.throws(
      () => addon.person_greet({}),
      /failed to downcast.*JsBox.*Person/
    );
  });

  it("should type check dynamic type", function () {
    const unit = addon.external_unit();

    assert.throws(() => addon.person_greet(unit), /failed to downcast/);
  });

  it("should be able to wrap a Rust value in an object", () => {
    const msg = "Hello, World!";
    const o = {};

    addon.wrapString(o, msg);
    assert.strictEqual(addon.unwrapString(o), msg);
  });

  it("should not be able to wrap an object twice", () => {
    const o = {};

    addon.wrapString(o, "Hello, World!");
    assert.throws(() => addon.wrapString(o, "nope"), /already wrapped/);
  });

  it("should not be able to unwrap an object that was not wrapped", () => {
    const o = {};

    assert.throws(() => addon.unwrapString(o), /not wrapped/);
  });
});

File: test/napi/src/js/boxed.rs
use std::cell::RefCell;

use neon::{prelude::*, types::extract::Boxed};

pub struct Person {
    name: String,
}

impl Finalize for Person {}

impl Person {
    fn new(name: impl ToString) -> Self {
        Self {
            name: name.to_string(),
        }
    }

    fn greet(&self) -> String {
        format!("Hello, {}!", self.name)
    }

    fn set_name(&mut self, name: impl ToString) {
        self.name = name.to_string();
    }
}

pub fn person_new(mut cx: FunctionContext) -> JsResult<JsBox<Person>> {
    let name = cx.argument::<JsString>(0)?.value(&mut cx);
    let person = Person::new(name);

    Ok(cx.boxed(person))
}

pub fn person_greet(mut cx: FunctionContext) -> JsResult<JsString> {
    let person = cx.argument::<JsBox<Person>>(0)?;
    let greeting = cx.string(person.greet());

    Ok(greeting)
}

pub fn ref_person_new(mut cx: FunctionContext) -> JsResult<JsValue> {
    let name = cx.argument::<JsString>(0)?.value(&mut cx);
    let person = RefCell::new(Person::new(name));

    Ok(cx.boxed(person).upcast())
}

pub fn ref_person_greet(mut cx: FunctionContext) -> JsResult<JsString> {
    let person = cx.argument::<JsBox<RefCell<Person>>>(0)?;
    let greeting = cx.string(person.borrow().greet());

    Ok(greeting)
}

pub fn ref_person_set_name(mut cx: FunctionContext) -> JsResult<JsUndefined> {
    let person = cx.argument::<JsBox<RefCell<Person>>>(0)?;
    let name = cx.argument::<JsString>(1)?.value(&mut cx);

    person.borrow_mut().set_name(name);

    Ok(cx.undefined())
}

pub fn ref_person_fail(mut cx: FunctionContext) -> JsResult<JsUndefined> {
    let person = cx.argument::<JsBox<RefCell<Person>>>(0)?;
    let _borrow = person.borrow();
    let _borrow_mut = person.borrow_mut();

    Ok(cx.undefined())
}

pub fn external_unit(mut cx: FunctionContext) -> JsResult<JsBox<()>> {
    Ok(cx.boxed(()))
}

#[neon::export]
fn create_boxed_string(s: String) -> Boxed<String> {
    Boxed(s)
}

#[neon::export]
fn boxed_string_concat(Boxed(this): Boxed<String>, rhs: String) -> String {
    this + &rhs
}

#[neon::export]
// N.B.: Intentionally including unused `cx` and not using tuple struct pattern to test the macro
fn boxed_string_repeat(_cx: &mut FunctionContext, this: Boxed<String>, n: f64) -> String {
    this.0.repeat(n as usize)
}
